packages/@lwc/template-compiler/src/codegen/expression.ts

```diff
@@ -0,0 +1,169 @@
+                !(t.isMemberExpression(parent) && parent.property === node) &&
+                !(t.isProperty(parent) && parent.key === node) &&
+                !codeGen.isLocalIdentifier(node) &&
+                !expressionScopes.isScopedToExpression(node)
+            ) {
+                this.replace(t.memberExpression(t.identifier(TEMPLATE_PARAMS.INSTANCE), node));
+            }
+        },
+    });
+
+    return expression as t.Expression;
+}
+
+/**
+ * Track the variables that come in and out of scope in various parts of a
+ * template expression. Arrow functions can return arrow functions, which can lead to
+ * variable shadowing, which needs to be handled correctly.
+ */
+class ExpressionScopes {
+    variableShadowingCount = new Map<VariableName, VariableShadowingMultiplicity>();
+    arrowFnVariables = new Map<t.ArrowFunctionExpression, Set<VariableName>>();
+
+    enterScope(node: t.ArrowFunctionExpression) {
+        const variableNamesIntroduced: VariableNames = new Set();
+        for (const param of node.params) {
+            collectParams(param, variableNamesIntroduced);
+        }
+        for (const varName of variableNamesIntroduced) {
+            this.variableShadowingCount.set(
+                varName,
+                (this.variableShadowingCount.get(varName) ?? 0) + 1
+            );
+        }
+        this.arrowFnVariables.set(node, variableNamesIntroduced);
+    }
+
+    exitScope(node: t.ArrowFunctionExpression) {
+        const varNames = this.arrowFnVariables.get(node);
+        if (varNames) {
+            for (const varName of varNames) {
+                this.variableShadowingCount.set(
+                    varName,
+                    this.variableShadowingCount.get(varName)! - 1
+                );
+            }
+        }
+    }
+
+    // If a variable was introduced in an arrow function and is still
+    // in scope, return true. Otherwise, return false.
+    isScopedToExpression(node: t.Identifier): boolean {
+        return !!this.variableShadowingCount.get(node.name);
+    }
+}
+
+function collectParams(node: t.BaseNode, vars: VariableNames) {
+    if (node.type === 'Identifier') {
+        collectParamsFromIdentifier(node as t.Identifier, vars);
+    } else if (node.type === 'ObjectPattern') {
+        collectParamsFromObjectPattern(node as t.ObjectPattern, vars);
+    } else if (node.type === 'Property') {
+        collectParamsFromProperty(node as t.Property, vars);
+    } else if (node.type === 'ArrayPattern') {
+        collectParamsFromArrayPattern(node as t.ArrayPattern, vars);
+    } else if (node.type === 'RestElement') {
+        collectParamsFromRestElement(node as t.RestElement, vars);
+    } else if (node.type === 'AssignmentPattern') {
+        collectParamsFromAssignmentPattern(node as t.AssignmentPattern, vars);
+    } else if (node.type === 'MemberExpression') {
+        collectParamsFromMemberExpression(node as t.MemberExpression, vars);
```

Comment:
I think we often create helpers that assert the type of an object to avoid these `as` casts. E.g.

https://github.com/salesforce/lwc/blob/35d169f47a5f95196afd45400a0a3295ec5f1cfc/packages/%40lwc/template-compiler/src/shared/ast.ts#L467-L469